File: dynamic-programming/matrix-chain-multiplication/golang/bottom_up_matrix_chain_multiplication.go
package main

import (
	"fmt"
	"math"
)

func matrixChanOrder(p []int) ([][]int, [][]int) {
	results := make([][]int, len(p))
	thresholds := make([][]int, len(p))

	for i := 0; i < len(p); i++ {
		results[i] = make([]int, len(p))
		thresholds[i] = make([]int, len(p))
		for j := 0; j < len(p); j++ {
			if i == j {
				results[i][j] = 0
				thresholds[i][j] = i
			} else {
				results[i][j] = math.MaxInt64
			}
		}
	}

	for diff := 1; diff < len(p)-1; diff++ {
		for i, j := 0, diff; j < len(p)-1; j++ {
			result := math.MaxInt64
			for k := i; k < j; k++ {
				if result > results[i][k]+results[k+1][j]+p[i]*p[k+1]*p[j+1] {
					result = results[i][k] + results[k+1][j] + p[i]*p[k+1]*p[j+1]
					thresholds[i][j] = k
				}
			}

			results[i][j] = result

			i++
		}
	}

	return results, thresholds
}

func printOptimalStrategy(thresholds [][]int, i, j int) {
	if i == j {
		fmt.Printf("A%d", thresholds[i][j])
	} else {
		fmt.Print("(")
		printOptimalStrategy(thresholds, i, thresholds[i][j])
		printOptimalStrategy(thresholds, thresholds[i][j]+1, j)
		fmt.Print(")")
	}
}

func main() {
	p := []int{30, 35, 15, 5, 10, 20, 25}

	fmt.Print("  i |")
	for i := 0; i < len(p)-1; i++ {
		fmt.Printf("%7d|", i)
	}
	fmt.Print("\n")
	fmt.Print("A[i]|")
	for i := 0; i < len(p)-1; i++ {
		fmt.Printf("%3d*%3d|", p[i], p[i+1])
	}
	fmt.Print("\n")

	results, thresholds := matrixChanOrder(p)

	for diff := 0; diff < len(p)-1; diff++ {
		fmt.Printf("--- diff = %d ---\n", diff)

		for i, j := 0, diff; j < len(p)-1; j++ {
			fmt.Printf("i = %d, j = %d / result = %d ", i, j, results[i][j])

			if i != j {
				printOptimalStrategy(thresholds, i, j)
			}

			fmt.Print("\n")

			i++
		}
	}
}

File: dynamic-programming/matrix-chain-multiplication/python/bottom_up_matrix_chain_multiplication.py
import sys


def matrix_chain_order(p):
    results = [[sys.maxsize for _ in range(len(p) + 1)] for _ in range(len(p) + 1)]
    thresholds = [[0 for _ in range(len(p) + 1)] for _ in range(len(p) + 1)]

    for i in range(len(p) + 1):
        results[i][i] = 0
        thresholds[i][i] = i

    diff = 1
    while diff < len(p) - 1:
        i, j = 0, diff

        while j < len(p) - 1:
            result = sys.maxsize
            for k in range(i, j):
                if result > results[i][k] + results[k + 1][j] + p[i] * p[k + 1] * p[j + 1]:
                    result = results[i][k] + results[k + 1][j] + p[i] * p[k + 1] * p[j + 1]
                    thresholds[i][j] = k

            results[i][j] = result

            i += 1
            j += 1

        diff += 1

    return results, thresholds


def print_optimal_strategy(thresholds, i, j):
    if i == j:
        print(f"A{thresholds[i][j]}", end="")
    else:
        print("(", end="")
        print_optimal_strategy(thresholds, i, thresholds[i][j])
        print_optimal_strategy(thresholds, thresholds[i][j] + 1, j)
        print(")", end="")


if __name__ == "__main__":
    p = [30, 35, 15, 5, 10, 20, 25]

    print("  i |", end="")
    for i in range(len(p) - 1):
        print(f"{i:^7}|", end="")
    print("")
    print("A[i]|", end="")
    for i in range(len(p) - 1):
        print(f"{p[i]:3}*{p[i + 1]:3}|", end="")
    print("")

    results, thresholds = matrix_chain_order(p)

    diff = 0
    while diff < len(p) - 1:
        print(f"--- diff = {diff} ---")

        i, j = 0, diff
        while j < len(p) - 1:
            print(f"i = {i}, j = {j} / result = {results[i][j]} ", end="")

            if i != j:
                print_optimal_strategy(thresholds, i, j)

            print("")

            i += 1
            j += 1

        diff += 1
